mempool/src/tests/multi_node_test.rs

```diff
@@ -0,0 +1,559 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    core_mempool::{CoreMempool, TimelineState},
+    network::{MempoolNetworkEvents, MempoolNetworkSender, MempoolSyncMsg},
+    shared_mempool::{start_shared_mempool, types::SharedMempoolNotification},
+    tests::common::TestTransaction,
+};
+use channel::{
+    self, diem_channel,
+    diem_channel::{Receiver, Sender},
+    message_queues::QueueStyle,
+};
+
+use diem_config::{
+    config::{PeerNetworkId, PeerRole},
+    network_id::{NetworkContext, NodeNetworkId},
+};
+use diem_infallible::{Mutex, RwLock};
+use diem_types::{
+    transaction::{GovernanceRole, SignedTransaction},
+    PeerId,
+};
+use futures::{
+    channel::mpsc::{self, unbounded, UnboundedReceiver},
+    executor::block_on,
+    future::FutureExt,
+    StreamExt,
+};
+use netcore::transport::ConnectionOrigin;
+use network::{
+    peer_manager::{
+        conn_notifs_channel, ConnectionNotification, ConnectionRequestSender,
+        PeerManagerNotification, PeerManagerRequest, PeerManagerRequestSender,
+    },
+    protocols::network::{NetworkEvents, NewNetworkEvents, NewNetworkSender},
+    transport::ConnectionMetadata,
+    DisconnectReason, ProtocolId,
+};
+use rand::{rngs::StdRng, SeedableRng};
+use std::{
+    collections::{HashMap, HashSet},
+    hash::Hash,
+    sync::Arc,
+};
+use storage_interface::mock::MockDbReader;
+use tokio::runtime::{Builder, Runtime};
+use vm_validator::mocks::mock_vm_validator::MockVMValidator;
+
+use crate::tests::node::{full_node, validator, vfn, Node, NodeTrait};
+use diem_config::config::NodeConfig;
+
+type NetworkHandle = (
+    NodeNetworkId,
+    MempoolNetworkSender,
+    NetworkEvents<MempoolSyncMsg>,
+);
+
+/// A test harness representing a combined network of Nodes and the mempool interactions between them
+#[derive(Default)]
+struct TestHarness {
+    mempools: HashMap<PeerId, Arc<Mutex<CoreMempool>>>,
+    network_reqs_rxs:
+        HashMap<PeerId, diem_channel::Receiver<(PeerId, ProtocolId), PeerManagerRequest>>,
+    network_notifs_txs:
+        HashMap<PeerId, diem_channel::Sender<(PeerId, ProtocolId), PeerManagerNotification>>,
+    network_conn_event_notifs_txs: HashMap<PeerId, conn_notifs_channel::Sender>,
+    runtimes: HashMap<PeerId, Runtime>,
+    subscribers: HashMap<PeerId, UnboundedReceiver<SharedMempoolNotification>>,
+    /// A mapping of secondary `PeerId` on other network interfaces to the primary `PeerId`
+    primary_peer_ids: HashMap<PeerId, PeerId>,
```

Comment:
this looks like something that should be a struct in your testnode defined in the previous commit or just another struct

having so many hashmaps is jarring and always bothers me that it could result in exists in one but not another... sorta like having two arrays versus having an array with a tuple.

Reply:
Yeah, let's see what I can cook up about it.  I was avoiding changing too much, and then changed everything.